server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/SortedDocsProducer.java

```diff
@@ -0,0 +1,397 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.bucket.composite;
+
+import org.apache.lucene.document.IntPoint;
+import org.apache.lucene.document.LongPoint;
+import org.apache.lucene.index.IndexOptions;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.PointValues;
+import org.apache.lucene.index.PostingsEnum;
+import org.apache.lucene.index.Terms;
+import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.search.CollectionTerminatedException;
+import org.apache.lucene.search.DocIdSet;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.MatchAllDocsQuery;
+import org.apache.lucene.search.PointRangeQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.DocIdSetBuilder;
+import org.apache.lucene.util.StringHelper;
+import org.elasticsearch.index.mapper.DateFieldMapper;
+import org.elasticsearch.index.mapper.KeywordFieldMapper;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.NumberFieldMapper;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+
+import java.io.IOException;
+import java.util.function.LongUnaryOperator;
+import java.util.function.ToLongFunction;
+
+/**
+ * A producer that visits composite buckets in the order of the value indexed in the leading source of the composite
+ * definition. It can be used to control which documents should be collected to produce the top composite buckets
+ * without visiting all documents in an index.
+ */
+abstract class SortedDocsProducer {
+    protected final String field;
+    protected final CompositeValuesCollectorQueue queue;
+
+    private SortedDocsProducer(String field, CompositeValuesCollectorQueue queue) {
+        this.field = field;
+        this.queue = queue;
+    }
+
+    /**
+     * Visits all non-deleted documents in <code>iterator</code> and pass documents that contain competitive composite buckets
+     * to the provided <code>sub</code> collector.
+     * Returns true if the queue is full and the current <code>leadSourceBucket</code> did not produce any competitive
+     * composite buckets.
+     */
+    protected boolean processBucket(LeafReaderContext context, LeafBucketCollector sub,
+                                    DocIdSetIterator iterator, Comparable<?> leadSourceBucket) throws IOException {
+        final int[] topCompositeCollected = new int[1];
+        final boolean[] hasCollected = new boolean[1];
```

Comment:
why do these need to be arrays if they only contain a single element?

Reply:
Because the values are set in the anonymous class below, I find it nicer than using Atomic and I saw this pattern in other locations in the codebase.